Enable proxy configuation within bbr tasks

miclip opened the pull request:
* Added params for HTTPS_PROXY, HTTP_PROXY and a SET_NO_PROXY flag
* Logic to export `no_proxy` when SET_NO_PROXY has been explicitly set to true
* Updated README with example for SET_NO_PROXY

miclip committed:
Enable proxy configuation within bbr tasks (#3)

* Added params for HTTPS_PROXY, HTTP_PROXY and a SET_NO_PROXY flag
* Logic to export `no_proxy` when SET_NO_PROXY has been explicitly set to true
* Updated README with example for SET_NO_PROXY

miclip committed:
Enable proxy configuation within bbr tasks (#3)

* Added params for HTTPS_PROXY, HTTP_PROXY and a SET_NO_PROXY flag
* Logic to export `no_proxy` when SET_NO_PROXY has been explicitly set to true
* Updated README with example for SET_NO_PROXY

gmrodgers committed:
Check apply changes is not inflight before taking a backup

[#165442961]

Signed-off-by: Jake Klein <jklein@pivotal.io>

Commit:
Only message output to stdout is the apply changes message

[#165442961]

Signed-off-by: Glen Rodgers <grodgers@pivotal.io>

gmrodgers committed:
Fix typo

[#165442961]

Signed-off-by: Jake Klein <jklein@pivotal.io>

gmrodgers commented:
Hey @miclip, thanks for the PR. We've placed this in pre-IPM, ready to be scheduled.

alamages committed:
Merge pull request #54 from pivotal-cf/check-opsman-status

Check apply changes is not inflight before taking a backup

alamages committed:
README now documents check-opsman-status task

[#166612123]

alamages committed:
check-opsman-status is now the first task

[#166612123]

alamages committed:
Remove tag from check-opsman-status in pks example

[#166612123]

aclevername commented:
Hey @miclip , thanks for the PR.

After revieiwng it the changes look good, however I've noticed the changes have only been made to the `bbr-backup-pas`, `bbr-backup-pks-clusters`, `bbr-backup-pks`, `bbr-cleanup-pas`, `bbr-cleanup-pks-clusters`, `bbr-cleanup-pks`, `lock-pks` and `unlock-pks` but not to the other tasks that do http/s requests, `bbr-backup-ert`, `bbr-cleanup-ert`. these tasks are the same as the `*-pas` but are kept as a legacy to when PAS was called ERT (we haven't deleted them encase customers are still using them). Could you include these changes in those tasks as well please?

When we talked through the call we mentioned how we won't need these for `bbr-director-*` tasks because we just ssh to the directot directly and don't talk to the bosh api. However it looks like from the readme updated that you invision world where we might need these variables to be exported when talking to OPSMAN API- in which case all tasks that talk to the OPSMAN API will need these changes.  Can you update the `bbr-director-*` and the `export-om-installation` to support these environment varaibles them as well please?

Finally since this PR has been opened we have also added a new task `check-opsman-status` (see https://github.com/pivotal-cf/bbr-pcf-pipeline-tasks/pull/54) to master. This new task talks to the OPSMAN api, so will also need these changes.

Thanks!

miclip commented:
Sure, I can make that change. I also think HTTP_PROXY should be removed as that should be managed from the concourse deployment.

miclip committed:
Removed HTTP_PROXY and added SET_NO_PROXY to ert tasks

miclip committed:
Removed HTTP_PROXY and added SET_NO_PROXY to ert tasks

miclip commented:
I don't think we need to set `no_proxy` for just the opsman api as it's something that can be managed from the bosh deployment, as could the director ip to be honest. I feel that the director IP is different as it's not really known and typically isn't accessed directly. 

Some might argue that `no_proxy` shouldn't be set from the task at all and be exclusively managed from the concourse deployment either as a list of IP's or the `[1,2,3,4,5,5,6,7,8]` workaround to cover all IPs in which case this PR isn't needed.

aclevername commented on scripts/export-director-metadata:
```diff
@@ -51,6 +51,12 @@ if [ ! -z ${OPSMAN_PRIVATE_KEY:+x} ]; then
 
+# Set NO_PROXY for BOSH Director and opsman host
+if [ ! -z ${SET_NO_PROXY:+x} ] && [ $SET_NO_PROXY = true ]; then
+    export NO_PROXY="${BOSH_ENVIRONMENT},$(basename "${OPSMAN_URL}")"
```
Do we need `OPSMAN_URL`?

aclevername commented on README.md:
```diff
@@ -54,6 +54,26 @@ Running regular backups (at least every 24 hours) and storing multiple copies of
+
+BBR tasks for backing up deployments (pas, pks) leverage the BOSH API and will result in HTTP requests to the director. 
+
+Setting the `SET_NO_PROXY` parameter on the tasks for `bbr-backup-pas`, `bbr-backup-pks` and `bbr-backup-pas-clusters` will result in a `NO_PROXY` environment variable being exported that contains the BOSH Director IP and OpsManager hostname. 
```
Can you remove reference to adding OpsManager Host and update the list of tasks or remove the list completely

aclevername commented on README.md:
````diff
@@ -54,6 +54,26 @@ Running regular backups (at least every 24 hours) and storing multiple copies of
+    SET_NO_PROXY: true
+```
+
+`HTTPS_PROXY` and `HTTP_PROXY` parameters are also available to be set and passed through to the container if a proxy is required. 
````
This is now redundant

aclevername commented on scripts/export-director-metadata:
```diff
@@ -51,6 +51,12 @@ if [ ! -z ${OPSMAN_PRIVATE_KEY:+x} ]; then
 
+# Set NO_PROXY for BOSH Director and opsman host
+if [ ! -z ${SET_NO_PROXY:+x} ] && [ $SET_NO_PROXY = true ]; then
+    export NO_PROXY="${BOSH_ENVIRONMENT},$(basename "${OPSMAN_URL}")"
```
Can you remove the `OPSMAN_URL`

miclip replied:
Yes, I added the `NO_PROXY` so we don't override any existing values set in the bosh deployment.

aclevername commented on tasks/bbr-backup-ert/task.yml:
```diff
@@ -37,3 +37,9 @@ params:
+
+  # Config NO_PROXY for BOSH cli and OM
+  # Setting SET_NO_PROXY: true results in NO_PROXY being exported 
+  #   with the BOSH Director IP and OPSMAN host
```
Remove reference to OPSMAN host for all tasks

terminatingcode replied:
This would essentially be replacing the commented text with:
```
  # Config NO_PROXY for BOSH cli and OM
  # Setting SET_NO_PROXY: true results in NO_PROXY being exported 
  # with the BOSH Director IP
```

miclip replied:
Sorry, missed those references.

aclevername commented on scripts/export-director-metadata:
```diff
@@ -51,6 +51,12 @@ if [ ! -z ${OPSMAN_PRIVATE_KEY:+x} ]; then
   sleep 10s
 fi
 
+# Set NO_PROXY for BOSH Director and opsman host
```
Remove reference top opsman host

miclip committed:
Removed references to ops manaager URL
Removed list of tasks, added ert to description
Appending include existing NO_PROXY in export

miclip committed:
Removed references to ops manaager URL
Removed list of tasks, added ert to description
Appending include existing NO_PROXY in export

terminatingcode commented on scripts/export-director-metadata:
```diff
@@ -51,6 +51,12 @@ if [ ! -z ${OPSMAN_PRIVATE_KEY:+x} ]; then
 
+# Set NO_PROXY for BOSH Director
+if [ ! -z ${SET_NO_PROXY:+x} ] && [ $SET_NO_PROXY = true ]; then
+    export NO_PROXY="$NO_PROXY,${BOSH_ENVIRONMENT}")"
```
Hi @miclip!

Looks like we have a failure here due to a typo. Might be leftover from the `basename` from  before:
```
bbr-pipeline-tasks-repo/tasks/bbr-backup-pas/../../scripts/export-director-metadata: line 56: syntax error near unexpected token `)'
```

miclip replied:
Thanks, weird that our test worked. I guess no proxy isn't that particular

terminatingcode reviewed (changes requested):
Small changes only.

miclip committed:
Corrected comments and syntax issue

miclip committed:
Corrected comments and syntax issue

miclip committed:
Merge branch 'master' of github.com:miclip/bbr-pcf-pipeline-tasks

terminatingcode commented:
> I don't think we need to set `no_proxy` for just the opsman api as it's something that can be managed from the bosh deployment, as could the director ip to be honest. I feel that the director IP is different as it's not really known and typically isn't accessed directly.
> 
> Some might argue that `no_proxy` shouldn't be set from the task at all and be exclusively managed from the concourse deployment either as a list of IP's or the `[1,2,3,4,5,5,6,7,8]` workaround to cover all IPs in which case this PR isn't needed.

Sorry for the delay in following up on this discussion point, as you mentioned, do you believe it makes more sense for this to be configured on the deployment instead of the task? What is more convenient for the customer and their use case?

If it lives at the deployment level, the information about proxies do not leak into the bbr task.

Thoughts?

miclip commented:
I think it makes sense for the opsman url to be managed at the deployment level as that is a well-known endpoint. The bosh director ip isn't and in theory can change although probably doesn't. You can set `no_proxy` to `0,1,2,3,4,5,6,7,8,9` in the deployment which covers all IP addresses which would address the issue as well. 

It still feels like a good feature for the bbr tasks as it does discover the bosh director ip, a thumbs up for ease of use but I also could be convinced that it should be left to the deployment.

terminatingcode commented:
Thanks @miclip 
I think that's a good argument. Once it goes through the pipeline successfully again, we'll merge it 👍

terminatingcode commented:
Looking good @miclip 👍 
Thanks for taking the time and being so responsive to questions and changes!
